src/hotspot/share/oops/trainingData.cpp

```diff
@@ -0,0 +1,815 @@
+ * published by the Free Software Foundation.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ *
+ */
+
+#include "ci/ciEnv.hpp"
+#include "ci/ciMetadata.hpp"
+#include "cds/cdsConfig.hpp"
+#include "cds/metaspaceShared.hpp"
+#include "classfile/classLoaderData.hpp"
+#include "classfile/compactHashtable.hpp"
+#include "classfile/javaClasses.hpp"
+#include "classfile/symbolTable.hpp"
+#include "classfile/systemDictionaryShared.hpp"
+#include "compiler/compileTask.hpp"
+#include "memory/metadataFactory.hpp"
+#include "memory/metaspaceClosure.hpp"
+#include "memory/resourceArea.hpp"
+#include "oops/method.hpp"
+#include "oops/methodCounters.hpp"
+#include "oops/trainingData.hpp"
+#include "runtime/arguments.hpp"
+#include "runtime/javaThread.inline.hpp"
+#include "runtime/jniHandles.inline.hpp"
+#include "utilities/growableArray.hpp"
+
+TrainingData::TrainingDataSet TrainingData::_training_data_set(1024, 0x3fffffff);
+TrainingData::TrainingDataDictionary TrainingData::_archived_training_data_dictionary;
+TrainingData::TrainingDataDictionary TrainingData::_archived_training_data_dictionary_for_dumping;
+TrainingData::DumptimeTrainingDataDictionary* TrainingData::_dumptime_training_data_dictionary = nullptr;
+int TrainingData::TrainingDataLocker::_lock_mode;
+volatile bool TrainingData::TrainingDataLocker::_snapshot = false;
+
+MethodTrainingData::MethodTrainingData() {
+  assert(CDSConfig::is_dumping_static_archive() || UseSharedSpaces, "only for CDS");
+}
+
+KlassTrainingData::KlassTrainingData() {
+  assert(CDSConfig::is_dumping_static_archive() || UseSharedSpaces, "only for CDS");
+}
+
+CompileTrainingData::CompileTrainingData() : _level(-1), _compile_id(-1) {
+  assert(CDSConfig::is_dumping_static_archive() || UseSharedSpaces, "only for CDS");
+}
+
+void TrainingData::initialize() {
+  // this is a nop if training modes are not enabled
+  if (have_data() || need_data()) {
+    // Data structures that we have do not currently support iterative training. So you cannot replay
+    // and train at the same time. Going forward we may want to adjust iteration/search to enable that.
+    guarantee(have_data() != need_data(), "Iterative training is not supported");
+    TrainingDataLocker::initialize();
+  }
+}
+
+#if INCLUDE_CDS
```

Comment:
I think you need to exclude this file in `make/hotspot/lib/JvmFeatures.gmk` when CDS is not included in VM. So you don't need all these #if.

Reply:
But then I'd need to ifdef all the TD API usage, do you think that's better?

Reply:
You can use `NOT_CDS_RETURN*` macros in header file - you keep it, only .cpp is not included.